### PR TITLE
Add authentication methods to access private binary repository

### PR DESCRIPTION
**Pull request by mollyIV**
## Description

This pull request is introducing a support for accessing the binary dependencies form private repositories. 

## Testing

**Manual testing**
1. [Add a private binary dependency](https://github.com/Carthage/Carthage/blob/master/Documentation/Artifacts.md#binary-only-frameworks) to your `Cartfile`.
2. Run `carthage update --platform ios --use-netrc`.
3. Check if a private binary dependency has been downloaded.

**Unit tests**
The bunch of unit tests for `Netrc` implementation were implemented in `NetrcSpec.swift`. 

## Important ⚠️ 

The current implementation of `.netrc` parser does not support `default`, `account` and `macdef` parameters. Check [the documentation](https://www.gnu.org/software/inetutils/manual/html_node/The-_002enetrc-file.html) to get more details about given parameters. 

closes https://github.com/Carthage/Carthage/issues/1783

### CODE REVIEW

File: Source/CarthageKit/Netrc.swift [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    
+    enum NetrcError: Error {
+        case fileNotFound(String)
+        case unreadableFile(String)
+        case machineNotFound
+        case missingToken(String)
+        case missingValueForToken(String)
+    }
+    
+    static func load(from file: URL = URL(fileURLWithPath: "\(NSHomeDirectory())/.netrc")) throws -> [NetrcMachine] {
```
Comment (blender) [inline Source/CarthageKit/Netrc.swift]: Please use Result instead.  Do not throw.

Comment (mollyIV) [reply Source/CarthageKit/Netrc.swift]: Thanks for an advice, I replaced throwing errors with `Result` type 🙇

**Comment by hlineholm** [comment]
In order to test the parser I added a `~/.netrc` with an Artifactory machine and used it like this: [Add auth header if machine exists in .netrc](https://github.com/hlineholm/Carthage/commit/07c9b669cc498eae5396766446106ccb3496445c).

**Comment by mollyIV** [comment]
👋 I am working on adding an option to `update` command `useBinaryCredentialNetrc`. The usage could look like: 

```bash
carthage update platform iOS  --useBinaryCredentialNetrc
```

File: Source/CarthageKit/Project.swift [hunk 2]
```diff
@@ -230,15 +230,32 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+                    // `~/.netrc` file to determine authentication credentials
+                    switch Netrc.load() {
+                    case .success(let machines):
+                        if let index = machines.firstIndex(where: { $0.name == binary.url.host }) {
+                            let machine = machines[index]
+                            let authString = "\(machine.login):\(machine.password)"
+                            if let authData = authString.data(using: .utf8) {
+                                let authValue = "Basic \(authData.base64EncodedString())"
+                                request.addValue(authValue, forHTTPHeaderField: "Authorization");
+                            }
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: I'd take `Netrc` an actual type and then move this logic for finding the credentials into a method on it.

Comment (mollyIV) [reply Source/CarthageKit/Project.swift]: Thanks for the suggestion 💪 Just to double check that we are on the same page - basically are you thinking about adding a method to `Netrc` struct like:

```Swift
// Netrc.swift

public func credentials(for machine: String) -> NetrcCredentials? {
    ...
}
```

Example usage:

``Swift
switch Netrc.load() {
                    case .success(let netrc):
                        let credentials = netrc.credentials(for: binary.url.host)
                        request.addValue(credentials.authorization, forHTTPHeaderField: "Authorization");
```

WDYT? 🤔

Comment (mdiep) [reply Source/CarthageKit/Project.swift]: Yup 👍 

But:

1. I'd probably pass the entire `URL` as the input to the method
2. I might return the authorization value directly

```
case let .success(netrc):
  if let authorization = netrc.authorizatino(for: binary.url) {
    request.addValue(authorization, forHTTPHeaderField: "Authorization")
```

**Comment by hlineholm-flir** [comment]
@mollyIV why not use --n or --netrc similar to what _curl_ use?

File: Source/CarthageKit/Project.swift (hunk 2 above)
Comment (mollyIV) [reply Source/CarthageKit/Project.swift]: The changes have been implemented @mdiep 🙇

**Comment by hlineholm** [comment]
> 👋 I am working on adding an option to `update` command `useBinaryCredentialNetrc`. The usage could look like:
> 
> ```shell
> carthage update platform iOS  --useBinaryCredentialNetrc
> ```

In case you decide to use `--useBinaryCredentialNetrc`; it should be `--use-binary-credential-netrc` right?

**Comment by hasanPrestoq** [comment]
This capability would be very useful for me. It looks like it has been approved, and passed tests. Is there any possibility that it could be merged soon?

**Comment by mollyIV** [comment]
Hey @hasanPrestoq 👋 

The PR hasn't been approved yet 

image

however it is ready for another round - all the suggestions have been implemented 🙂

**Comment by nuno-vieira** [comment]
This would also be very useful to me. Will @mdiep and @blender re-review this anytime soon?

**Comment by RonanMcH** [comment]
We are also waiting on this. This would be a great help. @mollyIV great work!

**Comment by ScottPierce** [comment]
This would also be useful to my team.

File: Source/CarthageKit/Netrc.swift [hunk 3]
```diff
@@ -0,0 +1,75 @@
+import Foundation
+import Result
+
+internal struct NetrcMachine {
```
Comment (tmspzz) [inline Source/CarthageKit/Netrc.swift]: You don't need to mark this as internal. It's internal by default

Comment (mollyIV) [reply Source/CarthageKit/Netrc.swift]: Hey @tmspzz, thank you for your comments. I have implemented the changes, could you please have a look? 👀 🙇

**Comment by mollyIV** [comment]
Unfortunately the CI jobs have failed. I did a rebase on top of the `master` - did not help. Could you please help on that @tmspzz ? 🙏

**Comment by tmspzz** [comment]
This fixes it but it still needs a few approvals: https://github.com/Carthage/Carthage/pull/2823

**Comment by tmspzz** [comment]
As @jdhealy commented privately, it makes sense to add a CLI parameter to opt in to this since there are limitations on what is supported. 

Something like `--use-netrc` maybe

File: Source/carthage/Update.swift [hunk 4]
```diff
@@ -65,13 +68,15 @@ public struct UpdateCommand: CommandProtocol {
 			let buildDescription = "skip the building of dependencies after updating\n(ignored if --no-checkout option is present)"
 
 			let dependenciesUsage = "the dependency names to update, checkout and build"
+            let netrcOption = Option(key: "use-netrc", defaultValue: false, usage: "use authentication credentials from ~/.netrc file when downloading binary only frameworks")
```
Comment (mollyIV) [inline Source/carthage/Update.swift]: Hey @tmspzz 👋 

I added the `use-netrc` option according to your suggestion, however I am not sure if `update` command is the only one where we shall add the option 🤔

Thoughts?

File: Source/carthage/Update.swift [hunk 5]
```diff
@@ -90,7 +95,9 @@ public struct UpdateCommand: CommandProtocol {
 		return options.loadProject()
 			.flatMap(.merge) { project -> SignalProducer<(), CarthageError> in
 
-				let checkDependencies: SignalProducer<(), CarthageError>
+                project.useNetrc = options.useNetrc
```
Comment (mollyIV) [inline Source/carthage/Update.swift]: @tmspzz  I have been trying to modify `Project.func downloadBinaryFrameworkDefinition(binary: BinaryURL) -> SignalProducer<BinaryProject, CarthageError>` method by adding additional `useNetrc: Bool` param. However there are quite a few places where this method is used into and I was not able to get the `netrc` value from `update` command options in that cases.

File: Source/CarthageKit/Netrc.swift (hunk 3 above)
Comment (tmspzz) [reply Source/CarthageKit/Netrc.swift]: I think it should be added to `bootstrap`, `update` and possibly `build`

Comment (mollyIV) [reply Source/CarthageKit/Netrc.swift]: Looks like `bootstrap` command already shares the command options with `update` command: 

https://github.com/Carthage/Carthage/blob/68abf2d73791720457f5be1aab4b57df86483f88/Source/carthage/Bootstrap.swift#L12-L14

Added `--use-netrc` to `build` command 🙇

**Comment by mollyIV** [comment]
[Build logs](https://travis-ci.org/Carthage/Carthage/jobs/565069436) points that `RemoteVersionSpec` test has failed:

```
Test Suite 'RemoteVersionSpec' failed at 2019-07-29 17:11:22.644.
	 Executed 2 tests, with 1 failure (0 unexpected) in 0.604 (0.604) seconds
```

but only on:

image

I did run the test locally and it passed. Is this test flaky? 🤔

**Comment by tmspzz** [comment]
@mollyIV yes, that test is flaky

File: Source/carthage/Update.swift (hunk 5 above)
Comment (tmspzz) [reply Source/carthage/Update.swift]: Well, the flag should be passed down to the appropriate places all the way or defaulter to false if not relevant. Can you point me to where you're unable to access the flag?

Comment (mollyIV) [reply Source/carthage/Update.swift]: I just started by modifying a mentioned function to:

```
// Project.swift file

func downloadBinaryFrameworkDefinition(binary: BinaryURL, useNetrc: Bool) -> SignalProducer<BinaryProject, CarthageError>
```

and eventually the adding `netrc: Bool` parameter to 👇 became an issue:

```
// Project.swift file

func versions(for dependency: Dependency) -> SignalProducer<PinnedVersion, CarthageError>
```

That's why I decided to add a property to the `Project`.

**Comment by mollyIV** [comment]
> @mollyIV yes, that test is flaky

@tmspzz Could you please re-run the job to get the green status in that case? 🙏 🙇

File: Source/carthage/Update.swift (hunk 5 above)
Comment (mollyIV) [reply Source/carthage/Update.swift]: Hey @tmspzz 👋 

Just wanted to kindly ask if you maybe got a chance to have a look at this one? 🙏 Please feel free to ask more questions if something's not clear 😉 Thanks! 🙇

**Comment by NicholasTD07** [comment]
Thanks for the great work, team! Can't wait for this to get into Carthage so we can pull private binary repos.

Any updates on this?

**Comment by tmspzz** [comment]
I think this is almost ready to go, but I would like to see support for comments

**Comment by mollyIV** [comment]
> I think this is almost ready to go, but I would like to see support for comments

:wave: @tmspzz 
I implemented comments support for `.netrc` file parser, could you please have a look? 👀 
🙇

**Comment by tmspzz** [comment]
@mollyIV The looks good. I just want to check again why you had to add `useNetrc` to the `Project`

**Comment by werner77** [comment]
Just two comments regarding the implementation:
- You'll probably also want to apply the .netrc credentials for the urls specified in the binary definition, not only for the definition itself
- What about the `outdated` command? It makes sense for that command to also include the `--use-netrc` option.

File: Source/CarthageKit/Project.swift [hunk 6]
```diff
@@ -101,6 +101,10 @@ public final class Project { // swiftlint:disable:this type_body_length
 	/// Whether to use submodules for dependencies, or just check out their
 	/// working directories.
 	public var useSubmodules = false
+    
+    /// Wheter to use authentication credentials from ~/.netrc file
```
Comment (vmus) [inline Source/CarthageKit/Project.swift]: Original code is indented with tabs.

**Comment by mollyIV** [comment]
Hey @werner77 👋 

Thank you for your questions. 

> You'll probably also want to apply the .netrc credentials for the urls specified in the binary definition, not only for the definition itself

That's a very good point 👍 Using a `.netrc` file for both private binary definitions and binaries themselves makes sense 🙇 Most probably the place in the code we need to add it 👇 
 https://github.com/Carthage/Carthage/blob/c94228549c6c9dacee4b50a45f141e30de4bf7da/Source/CarthageKit/Project.swift#L1009-L1018

> What about the outdated command? It makes sense for that command to also include the --use-netrc option.

I guess it does make sense 👍 @tmspzz could you please confirm that? 🙏 

> What about OAuth2 authentication (Authorization: Bearer xxxxxxxxxx)?

The `curl` sets an authentication header as `Authorization: Basic xxxx...`. The solution provided in this PR is heavily inspired by it.

Can you please describe how we could support a `OAuth2 (Authorization: Bearer xxxxxxxxxx)` authentication? 🤔 If we decided to support it, it would be great to tackle it in a separate issue. 

🙇 🙇

**Comment by tmspzz** [comment]
> What about the outdated command? It makes sense for that command to also include the --use-netrc option.

> I guess it does make sense 👍 @tmspzz could you please confirm that? 🙏

Yep

**Comment by JUSTINMKAUFMAN** [comment]
Would be great to get this merged. @mollyIV has been graciously working on it for a while now and, in addition to the others who have spoken up, this would also be helpful to my team.

**Comment by liambutler-lawrence** [comment]
Awesome that this is already done. my team is trying to add Carthage support to a set of proprietary SDKs we distribute (via Artifactory) and not having a way to authenticate with Artifactory to pull the artifacts is a blocker for us. What needs to happen to get this merged?

**Comment by mollyIV** [comment]
Hey @liambutler-lawrence 👋 

> What needs to happen to get this merged?

I just implemented and pushed to the repo two missing parts: 
* Use credentials from `.netrc` file for binaries download if needed
* Add `use-netrc` option to `outdated` command

Now the MR is ready for the final (hopefully 😉) review. CC: @tmspzz 

Would be great though if anyone could help us to do some manual testing 🙏

**Comment by liambutler-lawrence** [comment]
@mollyIV This works for our setup! (JSON spec file is on a public GitHub repo, but the artifact is on Artifactory). Had to merge master into your branch to get `make install` to work, which I've committed to my own fork: https://github.com/liambutler-lawrence/Carthage. We're going to use this fork until this PR is merged.

**Comment by mofarajmandi** [comment]
fwiw, just tested using @liambutler-lawrence and it worked as expected i.e. downloaded private binary frameworks using `.netrc` thanks @mollyIV and @liambutler-lawrence !

**Comment by liambutler-lawrence** [comment]
Bump again. Would love to get back to using the mainline Carthage- while the fork works great internally, we don't want to require that our clients use a fork of Carthage, so we're still using ugly workarounds (like IP whitelisting) to distribute our SDKs publicly.

**Comment by nuno-vieira** [comment]
@tmspzz It looks to me that we have everything that we need to get this merged? Anything more missing?

**Comment by DavidBrunow** [comment]
The indentation does not match the existing codebase throughout the changes -- Carthage uses tabs instead of spaces so you might need to change your Xcode settings and reformat.

File: Source/CarthageKit/Project.swift (hunk 6 above)
Comment (DavidBrunow) [reply Source/CarthageKit/Project.swift]: Typo in the word "Whether"

File: Tests/CarthageKitTests/NetrcSpec.swift [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            
+            it("should throw error when machine parameter is missing") {
+                let content = "login anonymous password qwerty"
+                let error = Netrc.from(content).error
+                
+                switch error {
+                case .some(.machineNotFound):
+                    break
+                default:
+                    fail("Expected invalidJSON error")
```
Comment (DavidBrunow) [inline Tests/CarthageKitTests/NetrcSpec.swift]: These fail messages don't make sense to me -- is this parsing JSON?

**Comment by darrylbaylissav** [comment]
Hey,

We've kept an eye on this PR for a few months. excited to begin using the end result. We were wondering when to expect this to be merged, we're planning features based on it.

File: Source/CarthageKit/Project.swift (hunk 6 above)
Comment (mollyIV) [reply Source/CarthageKit/Project.swift]: Fixed ✔️

File: Tests/CarthageKitTests/NetrcSpec.swift (hunk 7 above)
Comment (mollyIV) [reply Tests/CarthageKitTests/NetrcSpec.swift]: Thanks for spotting that, a copy - paste mistake, fixed ✔️

**Comment by KyleLeneau** [comment]
@mdiep @tmspzz What is the chances of getting this merged into the mainline? I used the NSOperations fork of Carthage that has this feature and it works exactly how I would want it to work. I can now securely store pre-built binaries that I manage in our companies Artifactory instance (including the JSON file) and reference the versions in a Cartfile for the team.

I would really like to see this MR / Feature in the mainline soon so I can have the team use avoid the step of getting the NSOperations fork. Can you provide an update on this?

**Pull request merged**